token-swap/src/lib.rs

```diff
@@ -0,0 +1,863 @@
+    ///   3. `[writable]` Pool MINT account, $authority is the owner.
+    ///   4. `[writable]` token_a Account to withdraw FROM.
+    ///   5. `[writable]` token_b Account to withdraw FROM.
+    ///   6. `[writable]` token_a user Account.
+    ///   7. `[writable]` token_b user Account.
+    ///   userdata: SOURCE amount of pool tokens to transfer. User receives an output based on the
+    ///   percentage of the pool tokens that are returned.
+    Withdraw(u64),
+}
+
+/// Creates an 'Init' instruction
+pub fn init(
+    program_id: &Pubkey,
+    swap_pubkey: &Pubkey,
+    authority_pubkey: &Pubkey,
+    token_a_pubkey: &Pubkey,
+    token_b_pubkey: &Pubkey,
+    pool_pubkey: &Pubkey,
+    user_output_pubkey: &Pubkey,
+    fees: (u64, u64),
+) -> Result<Instruction, ProgramError> {
+    let data = SwapInstruction::Init(fees).serialize()?;
+
+    let accounts = vec![
+        AccountMeta::new(*swap_pubkey, true),
+        AccountMeta::new(*authority_pubkey, false),
+        AccountMeta::new(*token_a_pubkey, false),
+        AccountMeta::new(*token_b_pubkey, false),
+        AccountMeta::new(*pool_pubkey, false),
+        AccountMeta::new(*user_output_pubkey, false),
+    ];
+
+    Ok(Instruction {
+        program_id: *program_id,
+        accounts,
+        data,
+    })
+}
+
+pub fn unpack<T>(input: &[u8]) -> Result<&T, ProgramError> {
+    if input.len() < size_of::<u8>() + size_of::<T>() {
+        return Err(ProgramError::InvalidAccountData);
+    }
+    #[allow(clippy::cast_ptr_alignment)]
+    let val: &T = unsafe { &*(&input[1] as *const u8 as *const T) };
+    Ok(val)
+}
+
+impl SwapInstruction {
+    /// Deserializes a byte buffer into an [SwapInstruction](enum.SwapInstruction.html)
+    pub fn deserialize(input: &[u8]) -> Result<Self, ProgramError> {
+        if input.len() < size_of::<u8>() {
+            return Err(ProgramError::InvalidAccountData);
+        }
+        Ok(match input[0] {
+            0 => {
+                let fee: &(u64, u64) = unpack(input)?;
+                Self::Init(*fee)
+            }
+            1 => {
+                let fee: &u64 = unpack(input)?;
+                Self::Swap(*fee)
+            }
+            2 => {
+                let fee: &u64 = unpack(input)?;
+                Self::Deposit(*fee)
+            }
+            3 => {
+                let fee: &u64 = unpack(input)?;
+                Self::Withdraw(*fee)
+            }
+            _ => return Err(ProgramError::InvalidAccountData),
+        })
+    }
+
+    /// Serializes an [SwapInstruction](enum.SwapInstruction.html) into a byte buffer
+    pub fn serialize(self: &Self) -> Result<Vec<u8>, ProgramError> {
+        let mut output = vec![0u8; size_of::<SwapInstruction>()];
+        match self {
+            Self::Init(fees) => {
+                output[0] = 0;
+                #[allow(clippy::cast_ptr_alignment)]
+                let value = unsafe { &mut *(&mut output[1] as *mut u8 as *mut (u64, u64)) };
```

Comment:
Why not have a generic pack function like this (with proper bounds checks and maybe taking the id as argument and returning a vec):

```rust
pub fn pack<T>(input: T,  output: &mut Vec<u8>) {
    #[allow(clippy::cast_ptr_alignment)]
    let value = unsafe { &mut *(&mut output[1] as *mut u8 as *mut T) };
    *value = input;
}
```